rest/server_test.go

```diff
@@ -29,131 +30,196 @@ func (m *mockSecretService) WriteSecret(key string, value string, password strin
+			srv := NewSecretRestAPI(mockSecretService, serverPort)
+			defer mockSecretService.AssertExpectations(t)
+			serverCtx, serverCancel := context.WithCancel(context.Background())
+			defer serverCancel()
+			go func() {
+				srv.Start(serverCtx)
+			}()
+			time.Sleep(1 * time.Second)
+			req, _ := http.NewRequest(http.MethodGet, fmt.Sprintf("http://localhost:%s?getter=key", serverPort), nil)
+			req.Header.Set("X-Cipher", tt.mockInput.password)
+			client := http.Client{
+				Timeout: 5 * time.Second,
+			}
+			res, err := client.Do(req)
+			if tt.expectedError != nil {
+				require.EqualError(t, err, tt.expectedError.Error())
+			} else {
+				require.NoError(t, err)
+			}
+			responseData, _ := ioutil.ReadAll(res.Body)
+			require.Equal(t, tt.expectedBody, string(responseData))
+			require.Equal(t, tt.expectedCode, res.StatusCode)
+			log.Println("test case finalized" + tt.name)
+		})
 	}
-	res, err := client.Do(req)
-	responseData, _ := ioutil.ReadAll(res.Body)
-	require.Contains(t, string(responseData), "invalid character")
-	require.Equal(t, http.StatusInternalServerError, res.StatusCode)
-	require.NoError(t, err)
-
 }
-func TestSecretRestAPI_Post_FailedToWriteData(t *testing.T) {
-	serverPort := strconv.Itoa(rand.Intn(20000-10000) + 10000)
-	mockSecretService := newMockSecretService()
-	mockSecretService.On("WriteSecret", "key", "value", "password").Return(errors.New("Some Error"))
-	srv := NewSecretRestAPI(mockSecretService, serverPort)
-	serverCtx, serverCancel := context.WithCancel(context.Background())
-	defer serverCancel()
-	go func() {
-		srv.Start(serverCtx)
-	}()
-	time.Sleep(1 * time.Second)
-	jsonBody := []byte(`{"getter" : "key", "value" : "value"}`)
-	bodyReader := bytes.NewReader(jsonBody)
-
-	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("http://localhost:%s?getter=key", serverPort), bodyReader)
-	req.Header.Set("X-Cipher", "password")
 
-	client := http.Client{
-		Timeout: 5 * time.Second,
+func TestSecretRestAPI_Post(t *testing.T) {
+
+	var testCases = []struct {
+		name          string
+		value         string
+		key           string
+		password      string
+		jBody         string
+		expectedCode  int
+		expectedBody  string
+		expectedError error
+		isServiceCall bool
+	}{
+
+		{
+			name:          "successfully operates on sent data",
+			value:         "value",
+			key:           "key",
+			password:      "password",
+			jBody:         `{"getter" : "key", "value" : "value"}`,
+			expectedCode:  http.StatusCreated,
+			expectedError: nil,
+			expectedBody:  "",
+			isServiceCall: true,
+		},
+		{
+			name:          "Returns error when fails to process corrupted json",
+			value:         "value",
+			key:           "key",
+			password:      "password",
+			jBody:         `{sdasdadasdad""||";;;"}`,
+			expectedCode:  http.StatusBadRequest,
+			expectedError: nil,
+			expectedBody:  "invalid character",
```

Comment:
Still requires json format for responses for POST methods

Reply:
> Still requires json format for responses for POST methods

it has json format but the only way to handle all the possible errors is to have Contains method.